tensorboard/webapp/widgets/line_chart_v2/lib/drawable_test.ts

```diff
@@ -0,0 +1,261 @@
+import {DataSeries} from './types';
+
+class TestableDataDrawable extends DataDrawable {
+  renderFrame(): void {
+    const metadataMap = this.getMetadataMap();
+    for (const data of this.series) {
+      const metadata = metadataMap[data.id];
+      this.paintBrush.setLine(data.id, data.polyline, {
+        width: 1,
+        color: metadata.color,
+        visible: metadata.visible,
+      });
+    }
+  }
+  getSeriesData() {
+    return this.series;
+  }
+}
+
+function buildSeries(override: Partial<DataSeries>): DataSeries {
+  return {
+    id: 'foo',
+    points: [
+      {x: 0, y: -1},
+      {x: 1, y: -0.5},
+      {x: 2, y: 0},
+      {x: 3, y: 0.5},
+      {x: 4, y: 1},
+    ],
+    ...override,
+  };
+}
+
+describe('line_chart_v2/lib/drawable test', () => {
+  let option: DrawableConfig;
+  let root: TestableDataDrawable;
+  let renderFrameSpy: jasmine.Spy;
+  let svg: SVGElement;
+  let getMetadataMap: jasmine.Spy;
+
+  beforeEach(() => {
+    svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
+
+    getMetadataMap = jasmine.createSpy().and.returnValue({
+      foo: {
+        color: '#f00',
+        visible: true,
+      },
+    });
+
+    option = {
+      coordinator: new Coordinator(),
+      renderer: new SvgRenderer(svg),
+      getMetadataMap,
+    };
+    root = new TestableDataDrawable(option);
+    root.setData([buildSeries({id: 'foo'})]);
+    option.coordinator.setViewBoxRect({x: 0, y: -1, width: 5, height: 1});
+    root.setLayoutRect({x: 0, y: 0, width: 100, height: 100});
+    renderFrameSpy = spyOn(root, 'renderFrame');
+  });
+
+  describe('draw frame', () => {
+    it('invokes redraw on empty data', () => {
+      root.internalOnlyDrawFrame();
+
+      expect(renderFrameSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('does not re-render if paint is not dirty', () => {
+      root.internalOnlyDrawFrame();
+
+      expect(renderFrameSpy).toHaveBeenCalledTimes(1);
+
+      // Nothing changed for paint to be dirty.
+      root.internalOnlyDrawFrame();
+      expect(renderFrameSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('re-renders if explictly marked as dirty', () => {
+      root.internalOnlyDrawFrame();
+      root.markAsPaintDirty();
+      root.internalOnlyDrawFrame();
+
+      expect(renderFrameSpy).toHaveBeenCalledTimes(2);
+    });
+
+    // If the dimension of the DOM changes, even if the data has not changed, we need to
+    // repaint.
+    it('re-renders if layout has changed', () => {
+      root.internalOnlyDrawFrame();
+      root.setLayoutRect({x: 0, y: 0, width: 200, height: 200});
+
+      expect(renderFrameSpy).toHaveBeenCalledTimes(1);
+
+      root.internalOnlyDrawFrame();
+
+      expect(renderFrameSpy).toHaveBeenCalledTimes(2);
+    });
+
+    it('manages the DOM cache', () => {
+      // Use the real redraw and update SVG.
+      renderFrameSpy.and.callThrough();
+      getMetadataMap = getMetadataMap.and.returnValue({
```

Comment:
I presume this problem is about a readability, right?
I definitely can see why declaring `a`, `b`, and `c` is defined up in the closure and I can comprehend why this is a problem. If we are testing truly pure components, your way would be clean, but, for more complex stateful components, current pattern allows for composition that is a bit more powerful than setUp (you _can_ do composition with setUp but at that point, it is not that different from `beforeEach`). For instance,

```ts
describe('foo test', () => {
  beforeEach(() => {
    // global setUp
  });

  describe('feature A', () => {
    beforeEach(() => {
      // partial setUp or set some default values for mocks relevant for testing feature A
      // mock or setUp state a, b, c, d
    });

    describe('very complex thing that needs to be thoroughly tested', () => {
      beforeEach(() => {
        // more setUp for testing special case
        // update state c, d, e
      });

      it('case 1', () => {
        // change state a and e that are interesting for this spec
      })
      it('case 2', () => {...})
    });

    describe('another scenarios that require a lot of mocking', () => {
      beforeEach(() => {
        // more setUp for the second case
        // change b, d, e.
      });
    });
  });
});
```

I am not defending that this pattern is the best since there are numbers of times that I flipped through `beforeEach` which maybe the root of problem you are pointing at. However, given that other tests are written in mocha/jasmine way and given that composition definitely was nice, I am leaning towards not refactoring.

However, given that this component is not as stateful as Angular ones, I am willing to refactor this module into the pattern you are suggesting. Please let me know what you prefer.

Reply:
Indeed; that there can be multiple layers of `beforeEach` strikes me as
more of a drawback than a benefit. When I’m trying to understand what’s
going on in a specific test, that’s even more locations and even more
control flow and mutation that I have to keep in mind.

I guess I’m just not familiar with that pattern really being required.
I’ve done my fair share of Python/JS/Java testing for pure things, UI
things, and things that talk to external services, and have never needed
to do something like this. I just grepped the SourceCred codebase, for
instance; we only have a handful of `beforeEach`s, and they’re only used
to assert that `console.error` is not called in a block (which is an
acceptable use, in my eyes—they’re installing a test-agnostic spy, not
setting up test-specific state).

> Please let me know what you prefer.

Maybe: keep an open mind moving forward, but I won’t request changes
here if you’re leaning toward not refactoring it.

Reply:
I hope this does not invoke a round of code review but here is some example of larger test: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/webapp/runs/views/runs_table/runs_table_test.ts

This is not the best code, so outline of the organization of the code is:
- tests the table for rendering run names
- tests the table for sorting and other related features
- tests for hparams and metrics feature
- tests for hparams and its filtering capabilities

(with some hypothetical situations) complexities are:
- when testing for run/exp names, we want to have empty hparams and metrics <- achievable with `setUp`
- when testing for hparams, we want to populate more information <- also achievable with `setUp`
- when you want to put the component in certain state and just test for interaction with the filter, we render the component then click on some components to put the component in the testable state <- achievable with `setUp` _and_ some other utility function.

Composing with function definitely is nice and now I am starting to question why these BDD frameworks decided to take more hierarchical approach to test authoring... 🤔

Reply:
> I hope this does not invoke a round of code review

Granted. :-)

What you write makes sense to me. I honestly didn’t realize that the
`let` style was part of the BDD way. I thought it was just about
defining things in terms of behaviors and with natural-language spec
names, which seemed pretty reasonable to me. Never looked into it
further than that, really.

Reply:
> [...]to me. I honestly didn’t realize that the `let` style was part of the BDD..

I am not sure if it is a BDD thing but it is the pattern that people use with Jasmine (from the code I saw internally). With Mocha, the context, `this`, are the same through `before`, `beforeEach`, `it`, and `afterEach` and people tend to tack on stuff onto that context. This is terrible for typing but it used to be popular in JS era.